### PR TITLE
[contracts] enforce ordering in deploy script

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**
- make the deployment of library-consuming contracts (e.g. `ConditionalTransaction`) dependent on the linkage of dependency libraries (e.g. `Transfer`)
- create a list of contracts and dependencies
- link contracts in separate steps to work around a bug (https://github.com/trufflesuite/truffle/pull/1489)

### CODE REVIEW

File: packages/contracts/migrations/2_deploy_contracts.js [hunk 1]
```diff
@@ -14,75 +14,45 @@ const VirtualAppAgreement = artifacts.require("VirtualAppAgreement");
...
+  deployer.then(async () => {
+
+    const artifacts = [
+      {
+        'artifact': Transfer,
+        'dependents': [VirtualAppAgreement, ConditionalTransaction]
+      },
+      {
+        'artifact': StaticCall,
+        'dependents': [ConditionalTransaction, Conditional]
```
Comment (mvanderh) [inline packages/contracts/migrations/2_deploy_contracts.js]: Don't need quotes

File: packages/contracts/migrations/2_deploy_contracts.js [hunk 2]
```diff
@@ -14,75 +14,45 @@ const VirtualAppAgreement = artifacts.require("VirtualAppAgreement");
 const ETHBalanceRefundApp = artifacts.require("ETHBalanceRefundApp");
 
 module.exports = (deployer, network) => {
-  deployer.deploy(Transfer).then(instance => {
-    deployer.link(Transfer, [VirtualAppAgreement, ConditionalTransaction]);
-    if (!tdr.isDryRunNetworkName(network)) {
-      return tdr.appendInstance(instance);
+  deployer.then(async () => {
+
+    const artifacts = [
```
Comment (mvanderh) [inline packages/contracts/migrations/2_deploy_contracts.js]: Move this outside the function scope and call it ARTIFACTS instead.

File: packages/contracts/migrations/2_deploy_contracts.js [hunk 3]
```diff
@@ -14,75 +14,45 @@ const VirtualAppAgreement = artifacts.require("VirtualAppAgreement");
...
+      Conditional
+    ];
+
+    for (const entry of artifacts) {
+      const artifact = entry.artifact || entry;
+      const instance = await deployer.deploy(artifact);
+
+      // link each dependent contract individually
+      // https://github.com/trufflesuite/truffle/pull/1489
+      for (const dependent of entry.dependents || []) {
```
Comment (mvanderh) [inline packages/contracts/migrations/2_deploy_contracts.js]:
```suggestion
      if (entry.dependents) {
          for (const dependent of entry.dependents) {
```
This is just my style preference, makes this code easier to skim.

**Comment by snario** [comment]
Isn't the more common approach just to make multiple migration files? @IIIIllllIIIIllllIIIIllllIIIIllllIIIIll

**Comment by snario** [comment]
You could also use `deployer.then` https://truffleframework.com/docs/truffle/getting-started/running-migrations#deployer-then-function-

**Comment by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll** [comment]
I do use `deployer.then`!

**Pull request merged**